src/ansi-c/c_typecheck_code.cpp

```diff
@@ -909,44 +909,103 @@ void c_typecheck_baset::typecheck_spec_assigns_target(exprt &target)
   typecheck_expr(target);
 
   // fatal errors
-  bool must_throw = false;
   if(target.type().id() == ID_empty)
   {
-    must_throw = true;
     error().source_location = target.source_location();
     error() << "void-typed expressions not allowed as assigns clause targets"
             << eom;
+    throw 0;
   }
 
-  if(!target.get_bool(ID_C_lvalue) && target.id() != ID_pointer_object)
+  if(target.get_bool(ID_C_lvalue))
   {
-    must_throw = true;
-    error().source_location = target.source_location();
-    error() << "assigns clause target must be an lvalue or " CPROVER_PREFIX
-               "POINTER_OBJECT expression"
-            << eom;
+    if(has_subexpr(target, ID_side_effect))
+    {
+      error().source_location = target.source_location();
+      error() << "side-effects not allowed in assigns clause targets" << eom;
+      throw 0;
+    }
+    if(has_subexpr(target, ID_if))
+    {
+      error().source_location = target.source_location();
+      error() << "ternary expressions not allowed in assigns "
+                 "clause targets"
+              << eom;
+      throw 0;
+    }
```

Comment:
This repeats multiple times and might be worth factoring out into a `static` function.

Reply:
factored it using a local lambda